src/merklized_state/data_source.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the HotShot Query Service library.
+//
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not,
+// see <https://www.gnu.org/licenses/>.
+
+//! Data for the [`merklized_state`](super) API.
+//!
+//! This module facilitates storing the state of a Merkle Tree at a specific point in time
+//! and provides methods for querying and reconstructing the snapshot.
+//!
+use ark_serialize::CanonicalDeserialize;
+use derive_more::{Display, From};
+
+use async_trait::async_trait;
+use commit::Commitment;
+use derivative::Derivative;
+use hotshot::traits::ValidatedState;
+use hotshot_types::{data::Leaf, traits::node_implementation::NodeType};
+use jf_primitives::{
+    circuit::merkle_tree::MembershipProof,
+    merkle_tree::{
+        prelude::{MerkleNode, MerklePath},
+        Element, Index, MerkleTreeScheme, NodeValue, ToTraversalPath,
+    },
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+use serde_json::Value;
+use typenum::Unsigned;
+
+use std::{cmp::Ordering, sync::Arc};
+
+// This trait defines methods that a data source should implement
+// It enables retrieval of the membership path for a leaf node, which can be used to reconstruct the Merkle tree state.
+#[async_trait]
+pub trait MerklizedStateDataSource<Types>
+where
+    Types: NodeType,
+{
+    type Error: std::error::Error + std::fmt::Debug + Send + Sync + 'static;
+    async fn get_path<
+        E: Element + Send + DeserializeOwned,
+        I: Index + Send + ToTraversalPath<A> + DeserializeOwned,
+        A: Unsigned,
+        T: NodeValue + Send + CanonicalDeserialize,
+    >(
+        &self,
+        _state_type: &'static str,
+        _tree_height: usize,
+        _header_state_commitment_field: &'static str,
+        _snapshot: Snapshot<Types>,
+        _key: Value,
+    ) -> Result<MerklePath<E, I, T>, Self::Error> {
+        Ok(vec![MerkleNode::Empty])
+    }
+}
+
+// This trait defines methods for updating the storage with the merkle tree state.
+#[async_trait]
+pub trait UpdateStateData: Send + Sync {
```

Comment:
Same here, this should be generic on `State: MerklizedState` and the relevant parameters should be removed from `insert_merkle_nodes`.